File: Unity3D/Assets/BizX/Game/GenreGameSystems.cs
using System;
using UnityEngine;

namespace BizX.Game {
    [Serializable] public class VehicleState { public float Speed, Grip = 1f, Fuel = 100f, Durability = 100f, Heat; }
    [Serializable] public class FighterState { public int Health = 100, Frame, Combo; public float Stamina = 100f, Meter; }

    public sealed class GenreGameSystems : MonoBehaviour {
        public VehicleState Vehicle = new();
        public FighterState Fighter = new();
        public int Wanted;

        public void Drive(float throttle, float brake, float steering, float dt = 1f / 60f) {
            throttle = Mathf.Clamp01(throttle); brake = Mathf.Clamp01(brake);
            Vehicle.Speed = Mathf.Max(0f, Vehicle.Speed + (throttle * 18f - brake * 28f) * dt);
            Vehicle.Speed *= Mathf.Max(0f, 1f - Mathf.Abs(steering) * (1f - Vehicle.Grip) * dt);
            Vehicle.Fuel = Mathf.Max(0f, Vehicle.Fuel - throttle * .02f);
        }

        public void SetWanted(int delta) => Wanted = Mathf.Clamp(Wanted + delta, 0, 5);

        public int Attack(string action, bool blocked = false) {
            Fighter.Frame++;
            int damage = action switch { "light" => 6, "heavy" => 12, "special" => 20, _ => 0 };
            if (damage > 0) Fighter.Meter = Mathf.Min(100f, Fighter.Meter + damage * .5f);
            int finalDamage = blocked ? Mathf.Max(1, damage / 4) : damage;
            Fighter.Health = Mathf.Max(0, Fighter.Health - finalDamage);
            Fighter.Combo = blocked ? 0 : Fighter.Combo + (damage > 0 ? 1 : 0);
            return finalDamage;
        }
    }
}

File: Unreal5/BizXUnreal/Source/BizXUnreal/GenreGameSystems.cpp
#include "GenreGameSystems.h"
#include "Math/UnrealMathUtility.h"

void UGenreGameSystems::Drive(float Throttle, float Brake, float Steering, float Dt) {
    Throttle=FMath::Clamp(Throttle,0.f,1.f); Brake=FMath::Clamp(Brake,0.f,1.f);
    Vehicle.Speed=FMath::Max(0.f,Vehicle.Speed+(Throttle*18.f-Brake*28.f)*Dt);
    Vehicle.Speed*=FMath::Max(0.f,1.f-FMath::Abs(Steering)*(1.f-Vehicle.Grip)*Dt);
    Vehicle.Fuel=FMath::Max(0.f,Vehicle.Fuel-Throttle*.02f);
}
void UGenreGameSystems::SetWanted(int32 Delta){ Wanted=FMath::Clamp(Wanted+Delta,0,5); }
int32 UGenreGameSystems::Attack(FName Action,bool Blocked){
    Fighter.Frame++;
    int32 Damage=0;
    if(Action==TEXT("light")) Damage=6; else if(Action==TEXT("heavy")) Damage=12; else if(Action==TEXT("special")) Damage=20;
    if(Damage>0) Fighter.Meter=FMath::Min(100.f,Fighter.Meter+Damage*.5f);
    const int32 FinalDamage=Blocked?FMath::Max(1,Damage/4):Damage;
    Fighter.Health=FMath::Max(0,Fighter.Health-FinalDamage);
    Fighter.Combo=Blocked?0:(Fighter.Combo+(Damage>0?1:0));
    return FinalDamage;
}
